### PR TITLE
444 Add description hash in invoice

### PR DESCRIPTION
**Pull request by txalkan**
Adds support required for the async payments flow.

See the blueprint for details:

[2026-04-10__rln-async-payments__blueprint.pdf](https://github.com/user-attachments/files/26642164/2026-04-10__rln-async-payments__blueprint.pdf)

### CODE REVIEW

File: src/routes.rs [hunk 1]
```diff
@@ -2950,9 +2952,16 @@ pub(crate) async fn ln_invoice(
...
+        let description = match &payload.description_hash {
+            Some(description_hash) => Bolt11InvoiceDescription::Hash(
+                validate_and_parse_description_hash(description_hash)?,
+            ),
+            None => Bolt11InvoiceDescription::Direct(Description::empty()),
+        };
 
         let invoice_params = Bolt11InvoiceParameters {
             amount_msats: payload.amt_msat,
+            description,
```
Comment (borismaxi) [inline src/routes.rs:2964]: This change wires `description_hash`, but it still does not expose the **CLTV** policy that the authoritative design needs for the async inbound leg. As written, the host can only get LDK's default final CLTV here and in [src/sdk/mod.rs#L3164-L3178](https://github.com/txalkan/rgb-lightning-node/blob/18074f7/src/sdk/mod.rs#L3164-L3178), not the explicit inbound policy the plan expects.

Relevant plan text:

> On the LNURL callback request:
> 1. allocate a fresh `payment_hash` and `index`,
> 2. call `create_inbound_payment_for_hash(payment_hash, Some(amount_msat), inbound_invoice_expiry_secs, Some(inbound_min_final_cltv))`,
> 3. build a BOLT11 invoice using:
>    - `payment_hash = allocated hash`,
>    - `payment_secret = returned secret`,
>    - `description_hash = sha256(raw LNURL metadata string)`,
>    - `expiry_time = inbound invoice expiry`,
>    - `min_final_cltv_expiry_delta = inbound invoice policy`,
>    - `basic_mpp` enabled.
>
> For this specific design:
> - inbound LNURL invoice wall-clock expiry `x_in`: `3600s`
> - client invoice wall-clock expiry `x_out`: `900s`
> - inbound final CLTV `c_in`: `144` blocks
> - client final CLTV `c_out`: `18` blocks

The relevant defaulting is in [UTEXO-Protocol/rust-lightning/lightning/src/ln/channelmanager.rs#L12971-L13014](https://github.com/UTEXO-Protocol/rust-lightning/blob/a6a68d0e0e9760bc8b539cc2488a4949b5d1302a/lightning/src/ln/channelmanager.rs#L12971-L13014) and [UTEXO-Protocol/rust-lightning/lightning/src/ln/channelmanager.rs#L12897-L12935](https://github.com/UTEXO-Protocol/rust-lightning/blob/a6a68d0e0e9760bc8b539cc2488a4949b5d1302a/lightning/src/ln/channelmanager.rs#L12897-L12935).

File: src/test/hodl_invoice.rs
Comment (borismaxi) [inline src/test/hodl_invoice.rs:107]: This test is asserting behavior that directly conflicts with the authoritative settlement model. After an inbound HTLC reaches `PaymentClaimable`, the design says the **remaining safety window is driven by `claim_deadline`, not by wall-clock invoice expiry**. Keeping this test as-is will lock in the wrong invariant.

Relevant plan text:

> Two different timers matter
> 1. **Invoice expiry (`x`)**
>    - wall-clock freshness,
>    - protects against stale invoices being started.
> 2. **CLTV / `claim_deadline`**
>    - actual settlement safety window once the payment is in-flight.
>
> The long window for an occasionally offline client must come mainly from the **inbound invoice's final CLTV delta**, not from a longer invoice expiry.
>
> Use the actual `claim_deadline` from `PaymentClaimable` as the authoritative inbound timer.

Relevant blueprint text:

> Happy path:
> RLN keeps the inbound HTLC claimable on PaymentClaimable.
> invoice-host requests the matching invoice from recipient-client, pays the outbound leg, then claims inbound using the revealed preimage.
>
> Timeouts and retries:
> Expiry only bounds the lifetime of a specific invoice attempt.

The runtime currently enforces the same non-compliant behavior in [src/ldk.rs#L1590-L1599](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/ldk.rs#L1590-L1599), [src/ldk.rs#L443-L487](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/ldk.rs#L443-L487), and [src/routes.rs#L1618-L1621](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/routes.rs#L1618-L1621).

File: src/test/hodl_invoice.rs [hunk 2]
```diff
@@ -477,10 +477,6 @@ async fn claim_hodl_invoice_btc_rgb() {
     .await;
 
     let _ = send_payment_with_status(node1_addr, invoice.clone(), HTLCStatus::Pending).await;
-    assert!(matches!(
-        invoice_status(node2_addr, &invoice).await,
-        InvoiceStatus::Pending
-    ));
```
Comment (borismaxi) [inline src/test/hodl_invoice.rs:483]: There should be explicit coverage for the `PaymentClaimable` validation rules here. The current handler in [src/ldk.rs#L1535-L1635](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/ldk.rs#L1535-L1635) ignores skimmed fees and only rejects `amount_msat < expected`, but the spec is stricter than that.

Relevant plan text:

> 3. **Inbound payment arrives**
> - LDK emits `PaymentClaimable`,
> - LSP verifies:
>   - `amount_msat == expected inbound amount`,
>   - `counterparty_skimmed_fee_msat == 0` unless explicitly allowed,
>   - `blocks_left = claim_deadline - current_height` is still safe,
>   - no earlier flow already succeeded for the same hash,
>   - if the same hash got paid again, fail the later payment.
>
> ### The LSP must reject ambiguous or lossy inbound payments
> - If the amount is not exact, reject.
> - If `counterparty_skimmed_fee_msat` is non-zero, reject unless you explicitly want this behavior.
> - If a second payment arrives for the same hash, fail it rather than accept a duplicate.

I would expect negative tests for overpayment, skimmed-fee delivery, and duplicate HTLC arrival before calling this implementation complete.

File: src/test/hodl_invoice.rs
Comment (borismaxi) [inline src/test/hodl_invoice.rs:473]: This only checks duplicate invoice creation, but the more dangerous duplicate case from the design is a second inbound HTLC for a `payment_hash` that already settled. The current `PaymentClaimable` path in [src/ldk.rs#L1572-L1635](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/ldk.rs#L1572-L1635) does not inspect existing settled state before upserting `Claimable`, so a duplicate HTLC can rewind a previously succeeded invoice instead of failing backwards.

Relevant plan text:

> 3. **Inbound payment arrives**
> - LDK emits `PaymentClaimable`,
> - LSP verifies:
>   - no earlier flow already succeeded for the same hash,
>   - if the same hash got paid again, fail the later payment.

Relevant blueprint text:

> Duplicate inbound HTLC policy:
> On the first valid PaymentClaimable for a new claim_id, transition that attempt to Claimable and continue the settlement flow.
>
> If the same claim_id or already-settled payment_hash is seen again, do not call `claim_funds` again; call `fail_htlc_backwards` and consider recording the attempt as a duplicate.

This needs a targeted duplicate-HTLC regression test, not just a duplicate-invoice-creation test.

Comment (borismaxi) [inline src/test/hodl_invoice.rs:508]: There should be a restart test around the first successful `/claimhodlinvoice` path here. The route flips state to `Claiming` in [src/routes.rs#L1624-L1629](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/routes.rs#L1624-L1629), but the preimage is only persisted later by `PaymentClaimed` in [src/ldk.rs#L1697-L1704](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/ldk.rs#L1697-L1704). That leaves a crash window where recovery has no durable preimage even though the flow already consumed it.

Relevant plan text:

> Persist these objects:
> - client account record,
> - hash inventory state,
> - payment flow state,
> - outstanding `request_invoice` RPC requests,
> - any learned preimage from a paid flow, especially any preimage not yet used to claim inbound and any preimage needed for recovery, audit, or accounting.
>
> On restart:
> 1. reload all persisted flows,
> 2. process persisted `PaymentClaimable` / `PaymentClaimed` events,
> 3. if a preimage was already learned but inbound not yet claimed, call `claim_funds` immediately if still before `claim_deadline`,
> 4. if a flow is waiting for client action and still safe, replay `request_invoice` on the next `open_session`,
> 5. otherwise mark the flow failed and burn the hash.

Without a restart regression here, this durability requirement is not actually validated.

File: src/routes.rs
Comment (borismaxi) [inline src/routes.rs:2257]: `invoice_status` is reading raw inbound state instead of the reconciled view, so it can report stale `Pending` or `Claimable` results until some other API runs expiry/deadline reconciliation first. The mismatch is between this route and [src/sdk/mod.rs#L3092-L3115](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/sdk/mod.rs#L3092-L3115) on one side and [src/ldk.rs#L426-L489](https://github.com/txalkan/rgb-lightning-node/blob/68486c9/src/ldk.rs#L426-L489) on the other.

Relevant plan text:

> ### `lnaddr.flow_status`
> Purpose:
> - reconcile after reconnect,
> - learn whether a given flow settled, failed, or is still waiting.
>
> Response:
> {
>   "flow_id": "9a3f...",
>   "state": "settled",
>   ...
> }

Even though this endpoint is not the final `lnaddr.flow_status` API, the broader design expectation is that status surfaces should reflect reconciled state, not whichever stale status happened to be persisted last.

**Review by borismaxi — COMMENTED** [review]
Found some issues. The most severe is the CTLV issue.

**Comment by txalkan** [comment]
> Found some issues. The most severe is the CTLV issue.

Hi Boris! Thanks for looking into it. This PR is only for the canonical description hash; the full design has not been implemented yet. Have you seen the utexo-lsp lightning address initial PR?

Looking forward to addressing your feedback in the next revision.

File: src/test/invoice.rs [hunk 3]
```diff
@@ -66,6 +71,43 @@ async fn invoice() {
...
+        .await
+        .unwrap();
+
+    let invoice = Bolt11Invoice::from_str(&res.invoice).unwrap();
+    assert!(matches!(
+        invoice.description(),
+        lightning_invoice::Bolt11InvoiceDescriptionRef::Hash(hash) if *hash == description_hash
+    ));
+}
+
```
Comment (borismaxi) [inline src/test/invoice.rs:110]: This positive-path test is the right core check for the scoped feature: it proves the HTTP API can accept a `description_hash` and emit an invoice whose description is encoded as a BOLT11 hash, not an inline description. I would add one more test next to it for an invalid `description_hash` string so the new validation path is covered as well, not just the happy path.

File: src/uniffi_api/tests.rs [hunk 4]
```diff
@@ -34,6 +34,7 @@ mod uniffi_smoke_tests {
             asset_id: None,
             asset_amount: None,
             payment_hash: None,
+            description_hash: None,
```
Comment (borismaxi) [inline src/uniffi_api/tests.rs:37]: The HTTP route is covered, but the scoped feature also changed the UniFFI surface. I would add one SDK / UniFFI-path test that sets `description_hash` and verifies the resulting invoice carries the hashed description tag. Right now the bindings compile path is updated, but there is no focused regression test proving this new field survives the non-HTTP call path end to end.

File: openapi.yaml [hunk 5]
```diff
@@ -710,7 +710,9 @@ paths:
       tags:
         - Invoices
       summary: Get a LN invoice
-      description: Get a LN invoice to receive a payment. Provide `payment_hash` to create a HODL invoice.
+      description: >-
+        Get a LN invoice to receive a payment. Provide `payment_hash` to create a HODL invoice.
+        Provide `description_hash` to include an out-of-band BOLT11 description hash.
```
Comment (borismaxi) [inline openapi.yaml:715]: The API description is technically correct, but I would make the scope more explicit in the PR narrative and maybe in the endpoint docs as well: this is support for setting the canonical BOLT11 `h` tag, and it is only one prerequisite for the larger async Lightning Address flow, not the async flow itself. That wording matters because the PR links the full blueprint, which otherwise invites reviewers to judge this small change as if it were implementing the whole design.

**Review by borismaxi — APPROVED** [review]
**LGTM**, pending minor nits.

Sorry, I was confused by linking to full `2026-04-10__rln-async-payments__blueprint.pdf` doc in this PR. My comments should be considered in a follow-up PR which implements the doc fully.

If this PR is scoped strictly to canonical BOLT11 `description_hash` plumbing, then the change looks good overall:

- the route and SDK accept `description_hash`,
- the hash is validated before invoice creation,
- the value is passed into `Bolt11InvoiceDescription::Hash(...)`,
- UniFFI / UDL / OpenAPI are updated consistently,
- there is a positive HTTP-path test confirming the emitted invoice uses the `h` tag.

I found some minor issues which are good to fix while we are in this PR.

I haven't seen https://github.com/UTEXO-Protocol/utexo-lsp/pull/4 yet.

**Comment by txalkan** [comment]
> **LGTM**, pending minor nits.

Thank you, Boris. I'm taking note of your feedback and resolving the conversation so we can move this forward.

**Review by gofman8 — APPROVED** [review]
